### PR TITLE
IEEE-35 add the EditTeamModal

### PR DESCRIPTION
**Pull request by Leo6Leo**
## Overview

- Resolves #IEEE-35
- Waiting to be updated

## Unit Tests Created

- In progress, will be updated later

## Steps to QA

- In progress, will be updated later

### CODE REVIEW

**Comment by Leo6Leo** [comment]
Encountered some issues: when try to import styles from .scss, it failed. Already compiled the scss file

**Comment by Leo6Leo** [comment]
CleanShot 2022-05-22 at 23 55 06@2x

And after I merge the most up-to-date develop branch, this error occured

File: hackathon_site/dashboard/frontend/src/pages/Orders/Orders.js [hunk 1]
```diff
@@ -2,14 +2,15 @@ import React from "react";
...
 import Typography from "@material-ui/core/Typography";
-
+import TeamModal from "../../components/dashboard/EditTeamModal/TeamModal";
 const Orders = () => {
     return (
-        <>
+        <div>
             <Header />
             <Typography variant="h1">Orders</Typography>
-            <p>IEEEEEE</p>
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/pages/Orders/Orders.js]: @Leo6Leo the test is failing because you changed the content on this page. Since the Edit Team Modal will be on the dashboard page you don't need to change this file.

**Comment by Leo6Leo** [comment]
I have only one problem left: how to test the clipboard. I have written the test down, but it seems like doesn't work. So I comment it out.

File: hackathon_site/dashboard/frontend/src/components/dashboard/EditTeamModal/EditTeamModal.tsx [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+import { addToCart, cartSelectors } from "slices/hardware/cartSlice";
+import { Category } from "api/types";
+import { Grid } from "@material-ui/core";
+//import styles from "../../inventory/ProductOverview/ProductOverview.module.scss";
+
+interface TeamModalProps {
+    teamCode: string;
+    canChangeTeam: boolean;
+}
+// export const ERROR_MESSAGES = {
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/dashboard/EditTeamModal/EditTeamModal.tsx]: Remove all the commented out code please :)

File: hackathon_site/dashboard/frontend/src/pages/Orders/Orders.js [hunk 3]
```diff
@@ -2,13 +2,14 @@ import React from "react";
 // import styles from "./Orders.module.scss";
 import Header from "components/general/Header/Header";
 import Typography from "@material-ui/core/Typography";
-
+import EditTeamModal from "../../components/dashboard/EditTeamModal/EditTeamModal";
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/pages/Orders/Orders.js]: can you revert this file to whats on develop, this page shouldn't have any changes

File: hackathon_site/dashboard/frontend/src/slices/ui/uiSlice.ts [hunk 4]
```diff
@@ -25,6 +26,7 @@ export const initialState: UIInitialState = {
         isCheckedOutTableVisible: true,
         isReturnedTableVisible: true,
         isPendingTableVisible: true,
+        isTeamModalVisible: true,
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/slices/ui/uiSlice.ts]: could you rename this to `isEditTeamVisible` because this isn't a modal, its technically a drawer

File: hackathon_site/dashboard/frontend/src/components/dashboard/EditTeamModal/EditTeamModal.tsx [hunk 5]
```diff
@@ -0,0 +1,478 @@
+import React from "react";
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/dashboard/EditTeamModal/EditTeamModal.tsx]: Could you rename this folder & its files to `EditTeamDrawer`, `EditTeamPanel` or even just `EditTeam` only because it's technically not a modal

File: hackathon_site/dashboard/frontend/src/slices/ui/uiSlice.ts (hunk 4 above)
Comment (YuYing-Liang) [reply hackathon_site/dashboard/frontend/src/slices/ui/uiSlice.ts]: Any why is this true by default? It should be false so the drawer is closed initially

File: hackathon_site/dashboard/frontend/src/slices/ui/uiSlice.ts [hunk 6]
```diff
@@ -54,6 +56,9 @@ const uiSlice = createSlice({
         closeProductOverview: (state: UIState) => {
             state.inventory.isProductOverviewVisible = false;
         },
+        removeTeamModalItem: (state: UIState) => {
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/slices/ui/uiSlice.ts]: I think you should rename this to closeEditTeam because "remove" is meant for deleting items

Also you need to have a reducer for opening Edit Team as well

File: hackathon_site/dashboard/frontend/src/components/dashboard/EditTeamModal/EditTeamModal.module.scss [hunk 7]
```diff
@@ -0,0 +1,84 @@
+@import "./../../../assets/abstracts/mixins";
+@import "./../../../assets/abstracts/variables";
+
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/dashboard/EditTeamModal/EditTeamModal.module.scss]: Could you add some padding here:
image

File: hackathon_site/dashboard/frontend/src/components/dashboard/EditTeamModal/EditTeamModal.module.scss [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+.heading {
+    margin-top: 20px;
+}
+
+.teamCode {
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/dashboard/EditTeamModal/EditTeamModal.module.scss]: you don't need this class to add a bold style, you can just use the `<strong>` or `<b>` tag

File: hackathon_site/dashboard/frontend/src/components/dashboard/EditTeamModal/EditTeamModal.module.scss [hunk 9]
```diff
@@ -0,0 +1,84 @@
+@import "./../../../assets/abstracts/mixins";
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/dashboard/EditTeamModal/EditTeamModal.module.scss]: You have a lot of unused classes, please go through and remove them from the file

**Review by YuYing-Liang — CHANGES_REQUESTED** [review]
Nice job for your first big UI ticket Leo!! Be proud :)
Could you move EditTeam component to display on the participant Dashboard page and have it open when you click "EDIT" on the team card.

**Comment by YuYing-Liang** [comment]
@Leo6Leo forgot to mention, you need to add some media queries to EditTeam so it displays properly on mobile, mui makes it easy to do this, in your current css class:

```
.editTeamModal {
   ...
   @include responsive(sm-down) {
       change the width to suit mobile screens here (width: 100% usually is the best)
       you can use md-down (medium-sized screens) and xs-down (extra small screens) too, 
      play around with the width and use chrome dev tools to make sure it's responsive
   }
}
```

File: hackathon_site/dashboard/frontend/src/slices/ui/uiSlice.ts (hunk 4 above)
Comment (Leo6Leo) [reply hackathon_site/dashboard/frontend/src/slices/ui/uiSlice.ts]: ahh I am just setting it as true for the testing purpose.

File: hackathon_site/dashboard/frontend/src/components/dashboard/EditTeamModal/EditTeamModal.module.scss (hunk 7 above)
Comment (Leo6Leo) [reply hackathon_site/dashboard/frontend/src/components/dashboard/EditTeamModal/EditTeamModal.module.scss]: do you mean to increase the distance between the change team form and the leave team button?

Comment (YuYing-Liang) [reply hackathon_site/dashboard/frontend/src/components/dashboard/EditTeamModal/EditTeamModal.module.scss]: Yes

Comment (Leo6Leo) [reply hackathon_site/dashboard/frontend/src/components/dashboard/EditTeamModal/EditTeamModal.module.scss]: CleanShot 2022-06-15 at 19 01 36@2x

But yuying I think it looks fine on my end. The distance between the leave team button and the team change form is quite large.

File: hackathon_site/dashboard/frontend/src/components/dashboard/EditTeamModal/EditTeamModal.tsx (hunk 5 above)
Comment (Leo6Leo) [reply hackathon_site/dashboard/frontend/src/components/dashboard/EditTeamModal/EditTeamModal.tsx]: CleanShot 2022-06-16 at 00 09 06@2x

On Jira, Graham set this name to EditTeamModal. Should I still change it?

**Comment by Leo6Leo** [comment]
CleanShot 2022-06-16 at 00 43 53@2x

My React tests passed, but it has these errors. Should I pay attention to them?

**Comment by YuYing-Liang** [comment]
> CleanShot 2022-06-16 at 00 43 53@2x
> 
> 
> 
> My React tests passed, but it has these errors. Should I pay attention to them?

Those are errors that will appear in the console when you deploy/develop your code. You should try to not have those errors. This error is saying the spacing prop is not being used on a Grid component container. So you either have to make that Grid a container or remove the spacing prop.

File: hackathon_site/dashboard/frontend/src/components/dashboard/EditTeamModal/EditTeamModal.module.scss (hunk 7 above)
Comment (YuYing-Liang) [reply hackathon_site/dashboard/frontend/src/components/dashboard/EditTeamModal/EditTeamModal.module.scss]: Yes I know, but you need to test your code on different screen widths and heights, on my screen it doesn't look like that.

File: hackathon_site/dashboard/frontend/src/components/dashboard/EditTeamModal/EditTeamModal.tsx (hunk 5 above)
Comment (YuYing-Liang) [reply hackathon_site/dashboard/frontend/src/components/dashboard/EditTeamModal/EditTeamModal.tsx]: Yeah, let's change it bc graham made that ticket really quickly and he's not on the team anymore

File: hackathon_site/dashboard/frontend/src/pages/Dashboard/Dashboard.tsx [hunk 10]
```diff
@@ -46,10 +60,20 @@ const Dashboard = () => {
...
 
+    const openEditTeamModal = () => {
+        //dispatch(getCurrentTeam());
+        dispatch(openTeamModalItem());
+    };
+
     return (
         <>
             <Header />
-            <ProductOverview showAddToCartButton={false} />
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/pages/Dashboard/Dashboard.tsx:52]: Why did you change this?

File: hackathon_site/dashboard/frontend/src/pages/Dashboard/Dashboard.tsx [hunk 11]
```diff
@@ -75,7 +99,7 @@ const Dashboard = () => {
                             key={0}
                             data-testid="team"
                         >
-                            <TeamCard handleEditTeam={() => alert("Editing Team")} />
+                            <TeamCard handleEditTeam={() => openEditTeamModal()} />
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/pages/Dashboard/Dashboard.tsx]: Because this function has no arguments you don't need to wrap a function around it 

`handleEditTeam={openEditTeamModal}` is sufficient

**Review by YuYing-Liang — COMMENTED** [review]
Left more comments, could you remove all the commented out code? It makes it easier to review.

For the clipboard test, you have to mock the function you call, probably with jest.spyOn() but I'll help you with this later, you can address the other comments for now.

File: hackathon_site/dashboard/frontend/src/pages/Dashboard/Dashboard.tsx (hunk 10 above)
Comment (Leo6Leo) [reply hackathon_site/dashboard/frontend/src/pages/Dashboard/Dashboard.tsx:52]: its a mistake, I have reverted it.

**Comment by Leo6Leo** [comment]
> Left more comments, could you remove all the commented out code? It makes it easier to review.
> 
> For the clipboard test, you have to mock the function you call, probably with jest.spyOn() but I'll help you with this later, you can address the other comments for now.

I have tried to do some research, and I found that most people use `import { mount } from "enzyme";` to mock the component, and find the button, and simulate onClick event. And then use jest.spyOn to capture the change in the clipboard. But I encountered some difficulty with using enzyme. Yeah, I might need your help with writing this clipboard test Yuying.

File: hackathon_site/dashboard/frontend/src/components/dashboard/EditTeam/EditTeam.tsx [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+import Alert from "@material-ui/lab/Alert";
+
+import FileCopyIcon from "@material-ui/icons/FileCopy";
+import SideSheetRight from "components/general/SideSheetRight/SideSheetRight";
+
+import styles from "./EditTeam.module.scss";
+import { isTeamModalVisibleSelector, closeTeamModalItem } from "slices/ui/uiSlice";
+import { useDispatch, useSelector } from "react-redux";
+import { Grid } from "@material-ui/core";
+import { maxTeamSize } from "../../../constants";
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/dashboard/EditTeam/EditTeam.tsx]: absolute import

File: hackathon_site/dashboard/frontend/src/components/dashboard/EditTeam/EditTeam.tsx [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+
+interface TeamModalProps {
+    teamCode: string;
+    canChangeTeam: boolean;
+    teamSize: number;
+}
+
+const TeamInfoBlock = ({ teamCode, canChangeTeam, teamSize }: TeamModalProps) => {
+    return (
+        <div>
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/components/dashboard/EditTeam/EditTeam.tsx]: you don't need this extra div here, and since you don't have any code before the return statement you can remove that too

```
const TeamInfoBlock = (...) => (
    <Alert ...>
         ....
    </Alert>
)
```

File: hackathon_site/dashboard/frontend/src/pages/Dashboard/Dashboard.tsx [hunk 14]
```diff
@@ -14,24 +14,38 @@ import Header from "components/general/Header/Header";
...
-import { getHardwareWithFilters, setFilters } from "slices/hardware/hardwareSlice";
+import {
+    getHardwareWithFilters,
+    getUpdatedHardwareDetails,
+    setFilters,
+} from "slices/hardware/hardwareSlice";
 import { getCategories } from "slices/hardware/categorySlice";
 import AlertBox from "components/general/AlertBox/AlertBox";
+import { openTeamModalItem } from "../../slices/ui/uiSlice";
+import EditTeam from "../../components/dashboard/EditTeam/EditTeam";
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/pages/Dashboard/Dashboard.tsx]: absolute imports

File: hackathon_site/dashboard/frontend/src/pages/Dashboard/Dashboard.tsx [hunk 15]
```diff
@@ -46,10 +60,20 @@ const Dashboard = () => {
         }
     }, [dispatch, hardwareInOrders]);
 
+    const openEditTeamModal = () => {
+        //dispatch(getCurrentTeam());
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/pages/Dashboard/Dashboard.tsx]: remove comment

Comment (YuYing-Liang) [reply hackathon_site/dashboard/frontend/src/pages/Dashboard/Dashboard.tsx]: also you don't need to brackets here
`const openEditTeamModal = () => dispatch(openTeamModalItem())`
is enough

File: hackathon_site/dashboard/frontend/src/pages/Dashboard/Dashboard.tsx [hunk 16]
```diff
@@ -46,10 +60,20 @@ const Dashboard = () => {
...
+        dispatch(openTeamModalItem());
+    };
+
     return (
         <>
             <Header />
-            <ProductOverview showAddToCartButton={false} />
+            <ProductOverview showAddToCartButton />
```
Comment (YuYing-Liang) [inline hackathon_site/dashboard/frontend/src/pages/Dashboard/Dashboard.tsx:68]: @Leo6Leo the props here are still wrong ^^

**Comment by YuYing-Liang** [comment]
> I have tried to do some research, and I found that most people use `import { mount } from "enzyme";` to mock the component, and find the button, and simulate onClick event. And then use jest.spyOn to capture the change in the clipboard. But I encountered some difficulty with using enzyme. Yeah, I might need your help with writing this clipboard test Yuying.

You would have difficulties with enzyme because we don't use it in our code lol. Enzyme is another testing library but we use React Testing Library instead. I'll play around with the test dw about it, i'll get back to you in the evening

File: hackathon_site/dashboard/frontend/src/pages/Dashboard/Dashboard.tsx (hunk 16 above)
Comment (Leo6Leo) [reply hackathon_site/dashboard/frontend/src/pages/Dashboard/Dashboard.tsx:68]: CleanShot 2022-06-17 at 09 59 03@2x

This is from develop branch, I didn't make any changes to this before. So What am I supposed to put here?

**Comment by YuYing-Liang** [comment]
@Leo6Leo try this for the copying to clipboard test

```
// put outside describe
const navigator = {
  clipboard: {
    writeText: (copyText: string) => jest.fn()
  },
};

// inside unit test
jest.spyOn(navigator.clipboard, "writeText");

< render edit team and click copy button >

expect(navigator.clipboard.writeText).toHaveBeenCalledWith(< whatever text was typed in >);
```

If that doesn't work, try some googling and if it's really annoying to implement, then dw about it. We only need to test things that could prevent a user from normally using the site.

**Comment by Leo6Leo** [comment]
CleanShot 2022-06-22 at 01 02 06@2x

Unfortunately, I have tried the test you gave to me, it still didn't work. Here is my code:

``` ts
test("Test the clipboard", async () => {
        // put outside describe
        const navigator = {
            clipboard: {
                writeText: (copyText: string) => jest.fn(),
            },
        };
        // inside unit test
        jest.spyOn(navigator.clipboard, "writeText");
        const store = makeStoreWithEntities({
            ui: {
                dashboard: {
                    isEditTeamVisible: true,
                },
            },
        });
        const { getByText } = render(
            <EditTeam teamCode={"ABCDE"} canChangeTeam={true} teamSize={1} />,
            {
                store,
            }
        );
        const copy_button = getByText("Copy");
        fireEvent.click(copy_button);
        expect(navigator.clipboard.writeText).toHaveBeenCalledWith("ABCDE");
});
```

**Comment by YuYing-Liang** [comment]
@Leo6Leo you're going to have to play around with it because I don't have a lot of time right now. You can look at the other jest.spyOn examples in our code (Ctrl+Shift+F "jest.spyOn" in pycharm) e.g. line 50 in withUserCheck.test.js to see if you can fix the problem. I might be able to take a look saturday or sunday.

**Comment by Leo6Leo** [comment]
> @Leo6Leo you're going to have to play around with it because I don't have a lot of time right now. You can look at the other jest.spyOn examples in our code (Ctrl+Shift+F "jest.spyOn" in pycharm) e.g. line 50 in withUserCheck.test.js to see if you can fix the problem. I might be able to take a look saturday or sunday.

Yes, I will play around with this thesedays.

**Review by YuYing-Liang — APPROVED** [review]
